crates/sui-core/src/authority/backpressure.rs

```diff
@@ -0,0 +1,315 @@
+    backpressure_sender: watch::Sender<bool>,
+}
+
+pub struct BackpressureSubscriber {
+    mgr: Arc<BackpressureManager>,
+}
+
+impl BackpressureManager {
+    pub fn new_for_tests() -> Arc<Self> {
+        Self::new_from_watermarks(Default::default())
+    }
+
+    fn new_from_watermarks(watermarks: Watermarks) -> Arc<Self> {
+        let (watermarks_sender, _) = watch::channel(watermarks);
+        let (backpressure_sender, _) = watch::channel(false);
+        Arc::new(Self {
+            watermarks_sender,
+            backpressure_sender,
+        })
+    }
+
+    pub fn new_from_checkpoint_store(store: &CheckpointStore) -> Arc<Self> {
+        let executed = store
+            .get_highest_executed_checkpoint_seq_number()
+            .expect("read cannot fail")
+            .unwrap_or_default();
+        let certified = store
+            .get_highest_synced_checkpoint_seq_number()
+            .expect("read cannot fail")
+            .unwrap_or_default();
+        info!(
+            ?executed,
+            ?certified,
+            "initializing backpressure manager from checkpoint store"
+        );
+        Self::new_from_watermarks(Watermarks {
+            executed,
+            certified,
+        })
+    }
+
+    pub fn update_highest_certified_checkpoint(&self, seq: CheckpointSequenceNumber) {
+        self.watermarks_sender.send_if_modified(|watermarks| {
+            if seq > watermarks.certified {
+                watermarks.certified = seq;
+                debug!(?watermarks, "updating highest certified checkpoint");
+                true
+            } else {
+                false
+            }
+        });
+    }
+
+    pub fn update_highest_executed_checkpoint(&self, seq: CheckpointSequenceNumber) {
+        self.watermarks_sender.send_if_modified(|watermarks| {
+            if seq > watermarks.executed {
+                debug_assert_eq!(seq, watermarks.executed + 1);
+                watermarks.executed = seq;
+                debug!(?watermarks, "updating highest executed checkpoint");
+                true
+            } else {
+                false
+            }
+        });
+    }
+
+    // Returns true if the backpressure state was changed.
+    pub fn set_backpressure(&self, backpressure: bool) -> bool {
+        self.backpressure_sender.send_if_modified(|bp| {
+            if *bp != backpressure {
+                debug!(?backpressure, "setting backpressure");
+                *bp = backpressure;
+                true
+            } else {
+                false
+            }
+        })
+    }
+
+    pub fn subscribe(self: &Arc<Self>) -> BackpressureSubscriber {
+        BackpressureSubscriber { mgr: self.clone() }
+    }
+}
+
+impl BackpressureSubscriber {
+    pub fn is_backpressure_active(&self) -> bool {
+        *self.mgr.backpressure_sender.borrow()
+    }
+
+    pub async fn await_backpressure(&self) {
```

Comment:
the name of this function to me sounds like it waits for backpressure to start rather than for it to finish

maybe `await_backpressure_end` or `await_backpressure_clearance` or `await_no_backpressure`?